// @angular/cli/commands/add-impl.js
"use strict";
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-global-tslint-disable no-any
const core_1 = require("@angular-devkit/core");
const tools_1 = require("@angular-devkit/schematics/tools");
const schematic_command_1 = require("../models/schematic-command");
const package_manager_1 = require("../utilities/package-manager");
class AddCommand extends schematic_command_1.SchematicCommand {
    constructor() {
        super(...arguments);
        this.allowPrivateSchematics = true;
    }
    async run(options) {
        if (!options.collection) {
            this.logger.fatal(`The "ng add" command requires a name argument to be specified eg. `
                + `${core_1.terminal.yellow('ng add [name] ')}. For more details, use "ng help".`);
            return 1;
        }
        const packageManager = package_manager_1.getPackageManager(this.workspace.root);
        const npmInstall = require('../tasks/npm-install').default;
        const packageName = options.collection.startsWith('@')
            ? options.collection.split('/', 2).join('/')
            : options.collection.split('/', 1)[0];
        // Remove the tag/version from the package name.
        const collectionName = (packageName.startsWith('@')
            ? packageName.split('@', 2).join('@')
            : packageName.split('@', 1).join('@')) + options.collection.slice(packageName.length);
        // We don't actually add the package to package.json, that would be the work of the package
        // itself.
        await npmInstall(packageName, this.logger, packageManager, this.workspace.root);
        const runOptions = {
            schematicOptions: options['--'] || [],
            workingDir: this.workspace.root,
            collectionName,
            schematicName: 'ng-add',
            allowPrivate: true,
            dryRun: false,
            force: false,
        };
        try {
            return await this.runSchematic(runOptions);
        }
        catch (e) {
            if (e instanceof tools_1.NodePackageDoesNotSupportSchematics) {
                this.logger.error(core_1.tags.oneLine `
          The package that you are trying to add does not support schematics. You can try using
          a different version of the package or contact the package author to add ng-add support.
        `);
                return 1;
            }
            throw e;
        }
    }
}
exports.AddCommand = AddCommand;
